extkeys/mnemonic.go

```diff
@@ -216,17 +232,61 @@ func (m *Mnemonic) ValidMnemonic(mnemonic string, language Language) bool {
 
 	// The number of words should be 12, 15, 18, 21 or 24
 	if numOfWords%3 != 0 || numOfWords < 12 || numOfWords > 24 {
-		return false
+		return errors.New("mnemonic contains an invalid number of words")
 	}
 
+	// Create reverse lookup map for dictionary
+	wordMap := map[string]int{}
+	for i, v := range wordList {
+		wordMap[v] = i
+	}
+
+	b := big.NewInt(0)
+
 	// Check if all words belong in the wordlist
 	for i := 0; i < numOfWords; i++ {
-		if !contains(wordList, words[i]) {
-			return false
+
+		wordIndex, found := wordMap[words[i]]
+		if !found {
+			return errors.New("not all words are found in dictionary")
 		}
+		var wordBytes [2]byte
+		binary.BigEndian.PutUint16(wordBytes[:], uint16(wordIndex))
+		b = b.Mul(b, rightShift11BitsDivider)
+		b = b.Or(b, big.NewInt(0).SetBytes(wordBytes[:]))
+	}
+
+	checksum := big.NewInt(0)
+	checksumMask := wordLengthChecksumMasksMapping[len(words)]
+	checksum = checksum.And(b, checksumMask)
+
+	b.Div(b, big.NewInt(0).Add(checksumMask, bigOne))
+
+	// Calculate entropy from mnemonic
+	entropy := b.Bytes()
+	entropy = padByteSlice(entropy, len(words)/3*4)
+
+	// Calculate checksum from entropy derived above
+	hasher := sha256.New()
+	hasher.Write(entropy)
+	computedChecksumBytes := hasher.Sum(nil)
+	computedChecksum := big.NewInt(int64(computedChecksumBytes[0]))
+
+	if l := len(words); l != 24 {
+		checksumShift := wordLengthChecksumShiftMapping[l]
+		computedChecksum.Div(computedChecksum, checksumShift)
 	}
 
-	return true
+	if checksum.Cmp(computedChecksum) != 0 {
+		return errors.New("checksum for mnemonic seed is invalid")
+	}
+
+	return nil
+}
+
+// ValidMnemonic validates mnemonic string
+func (m *Mnemonic) ValidMnemonic(mnemonic string, language Language) bool {
+	return m.ValidateMnemonic(mnemonic, language) == nil
```

Comment:
Needs to implement @adambabik's suggested change.

Reply:
This is done in latest commit. Let me know if it needs anything further.